### PR TITLE
[WIP] Install Cython in isolated env using fortunate-pkg

### LINKED ISSUE #3581: sdist build gets crashed under pip>=19 in dev mode
## Long story short

We have `cython` as an optional dependency. That's why we install it as a pre-requisite in the CI, as a separate step.
New pip creates a separate build virtualenv which doesn't have access to the place with cython installed which causes it to crash.

## Expected behaviour

It succeeds

## Actual behaviour

It tracebacks

## Steps to reproduce

* https://travis-ci.com/aio-libs/aiohttp/jobs/172249543#L198-L219
* https://ci.appveyor.com/project/aio-libs/aiohttp/build/job/uppd0qqw2sbisqtn#L46

## Your environment

Travis CI (doesn't really matter actually)

### PR DESCRIPTION
**Pull request by webknjaz**
The source code of "try pip install" is @ https://github.com/webknjaz/fortunate-pkg ~because I thought it might be useful for other projects~ because it's the only reliable and pip-implemented way of invoking PEP 517 build backend.

## What do these changes do?

This basically adds `fortunate-pkg` as a requisite of `build-system` (PEP 517 / PEP 518).

Then, `setup.py` uses it to trigger `pip install cython` and ignore the result. So it would install `Cython` into an isolated build env created by Pip only if the user's OS has the build toolchain available (GCC in particular) and if it succeeds to compile Cython.

## Are there changes in behavior for the user?

With modern pip they don't have to care about pre-installing cython req manually.

## Related issue number

This PoC should fix #3581 but needs futher discussion.
cc @pradyunsg plz participate

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by webknjaz** [comment]
P.S. Please don't merge this, let me do this once I'm sure it's fine.

**Comment by webknjaz** [comment]
FTR previously red tests got green but I still want to discuss this first.

**Comment by kxepal** [comment]
Honestly, this looks like a hack, a workwaround pip issue. Can we just wait while it will get fixed?

File: setup.py [hunk 1]
```diff
@@ -7,6 +7,13 @@
                               DistutilsPlatformError)
 
 from setuptools import Extension, setup
+try:
+    from fortunate_pkg import maybe_install_pkgs
+except ImportError:
+    import os
+    os.system('echo "Running outside of pip\'s isolated virtualenv."')
```
[the remark this answers is not in the document]
Comment (kxepal) [reply setup.py]: \+ many

[the remark this answers is not in the document]
Comment (webknjaz) [reply setup.py]: Because pip (or setuptools) will "eat" that. this is a hack to work around that.

[the remark this answers is not in the document]
Comment (webknjaz) [reply setup.py]: And even this is visible only if you pass `-v` to pip.

**Comment by webknjaz** [comment]
@kxepal pip has 3 months release process but this would require a lot of discussions so I can't even imagine how much time would it take.
Are you going to keep telling every single user to not update to pip 19 in issues?

File: setup.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (webknjaz) [reply setup.py]: I know that you probably think "Hold my beer, I'll go fix it in a few". But trust me, I'm deep down the rabbit hole already and there's no escape. That's why such hacks are needed.

**Comment by webknjaz** [comment]
Oh, and that might also need to go through the PEP process first to get a standard name for such optional requirement option + having "ignoring logic" requires some dynamic code which is not really a part of PEP 517 / PEP 518.

File: setup.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply setup.py]: can we at least use `subprocess.run`? `os.system` has been deprecated since about 2.6.

[the remark this answers is not in the document]
Comment (webknjaz) [reply setup.py]: right. yes, we can

**Comment by samuelcolvin** [comment]
makes sense, but can we add a note to `setup.py` explaining why this is necessary for anyone struggling in future, maybe including a link to this PR?

**Comment by webknjaz** [comment]
sure, good point

File: setup.py [hunk 2]
```diff
@@ -7,6 +7,13 @@
...
+else:
+    maybe_install_pkgs(('cython', ))
```
Comment (asvetlov) [inline setup.py]: We have `AIOHTTP_NO_EXTENSIONS` env var.
Skipping `cython` installation if the env var is set is a good idea maybe.
What do you think about?

Comment (asvetlov) [reply setup.py]: Nevermind

**Comment by webknjaz** [comment]
@asvetlov @samuelcolvin @kxepal

So I'm investigating ways of shipping the build script within this repo.
One way is to have an "installable dist" folder (with a separate `setup.py`) just for a builder tool and the other way is to use https://github.com/takluyver/intreehooks (pointed out by @pradyunsg) which allows working around the limitation of being unable to reference local scripts directly (https://discuss.python.org/t/pep-517-backend-bootstrapping/789).

So what do you think? Should I rely on intreehooks shim or hack it all within aiohttp repo?

**Comment by kxepal** [comment]
I think this is going too far and I feel that there is something wrong. Do we need to relay on PEP-517 at all if it requires so much work?

**Comment by webknjaz** [comment]
PEP 517 is fine, it just has certain limitations in its current state. Nothing is wrong with it and we'll have to rely on it anyway with pip 19+. We shouldn've done this a while back actually.

**Comment by webknjaz** [comment]
(Please do not merge this, I'll be doing further changes as needed)

**Comment by webknjaz** [comment]
@asvetlov most of this will be moved to a separate dist so the final diff will be way smaller.

**Comment by webknjaz** [comment]
@asvetlov @samuelcolvin @kxepal

This is now ready for further discussion.

**Comment by asvetlov** [comment]
I think the PR can be closed: we use an alternative solution for cython sources translation.

**Comment by webknjaz** [comment]
Okay, I can always restore this if needed.

**Pull request closed without merging**